### ISSUE TITLE
Invalid JWT signature created by PostgREST with binary jwt-secret

### ISSUE THREAD

**Issue by eric-brechemier** [begriffs/postgrest#495]
The verification on [jwt.io](http://jwt.io/) of the JSON Web Tokens generated by PostgREST fails when the configured `jwt-secret` is binary.

I am using latest stable version of PostgREST, `v0.3.0.4`, running on Mac OSX 10.11.1 El Capitan. I have the same issue on a server running Ubuntu 14.04 LTS.

Here are the steps to reproduce:

1. generate a binary secret in a file using openssl (only 8 bytes to keep it simple, but it is unsafe):
  ```sh
openssl rand 8 > jwt.secret
```

2. open the page [jwt.io](http://jwt.io/)

3. encode the binary secret as `base64url` for use in jwt.io:
   ```sh
base64UrlEncode()
{
  base64 |
  sed -e 's/+/-/g' -e 's#/#_#g' -e 's/=//g' # base64 to base64url
}
cat jwt.secret |
base64UrlEncode |
less
```

4. copy the JWT secret encoded in `base64url` format
(or use `pbcopy` instead of `less` at previous step)

5. in the page [jwt.io](http://jwt.io/), paste the JWT secret encoded in `base64url` format in the box labelled "VERIFY SIGNATURE" in "Decoded" section on the right of the page

6. under the input for the JWT secret, check the box ☑ `secret base64 encoded`

7. define a data type and a stored procedure to prepare a JSON Web Token for PostgREST (I have isolated them in a schema named `issue495`):
  ```sql
CREATE SCHEMA issue495;

CREATE TYPE issue495.jwt_claims 
AS ( role TEXT ) 
;

CREATE FUNCTION issue495.get_jwt_for_role(role TEXT)
RETURNS issue495.jwt_claims
RETURNS NULL ON NULL INPUT
LANGUAGE sql
IMMUTABLE
AS 'SELECT role AS role;'
;
```

8. start a new instance of PostgREST which connects to the schema `issue495` and makes the stored procedure available on port 4495 (for example) of localhost, using the binary `jwt-secret`:
  ```sh
dbuser="$(whoami)";
postgrest "postgres://$dbuser@localhost:5432/egull" \
  --anonymous "$dbuser" \
  --schema issue495 \
  --port 4495 \
  --jwt-secret "$(cat jwt.secret)"
```

9. POST a request to get the JSON Web Token for the role `issue495`:
  ```sh
curl -s -X POST \
  -H "Content-Type: application/json" \
  -d '{"role":"issue495"}'  \
  'http://localhost:4495/rpc/get_jwt_for_role' |
cut -d'"' -f4 | # cut the value of the "token" property
less
```

10. copy the JSON Web Token
(or use `pbcopy` instead of `less` at previous step)

11. paste the JSON Web Token in the input labelled "Encoded" on the left of the page

The result of the validation is displayed in a red bar below: "invalid signature".

**Comment by eric-brechemier** [comment]
In order to display the expected signature, modify the payload without changing the value, by removing a space character. The JSON Web Token is updated and the validation status below changes to "Signature Verified".

**Comment by eric-brechemier** [comment]
The expected signature can be generated from the shell, reading the secret into a parameter in the same way as it is provided to PostgREST, which rules out a corruption of the key by the command line:

```sh
# Parameters: role_name and secret
role_name="issue495"
secret="$(cat jwt.secret)"

base64UrlEncode()
{
  base64 |
  sed -e 's/+/-/g' -e 's#/#_#g' -e 's/=//g' # base64 to base64url
}

algorithm='"alg"'
hmac_sha256='"HS256"'
token_type='"typ"'
json_web_token='"JWT"'
header="{$algorithm:$hmac_sha256,$token_type:$json_web_token}"
base64url_header="$(printf '%s' "$header" | base64UrlEncode)"

role='"role"'
role_value="\"$role_name\""
payload="{$role:$role_value}"
base64url_payload="$(printf '%s' "$payload" | base64UrlEncode)"

message="$base64url_header.$base64url_payload"

# Note that the binary secret is provided through a parameter on the command line
signature="$(
  printf '%s' "$message" |
  openssl dgst -sha256 -hmac "$secret" -binary |
  base64UrlEncode
)"

# Result
jwt="$message.$signature"
echo "$jwt"
```

**Comment by begriffs** [comment]
Thanks for these great steps to reproduce. I'm seeing it locally as well. I wonder if it has to do with the way you're base64 encoding. For instance, I tried using the plain secret "abc123" and and jwt.io verified the signature. Then I did

```bash
$ echo "abc123" | base64UrlEncode
YWJjMTIzCg
```

Using that base64 value in jwt.io with the "secret base64 encoded" checkbox enabled now shows the same payload failing its signature verification.

**Comment by eric-brechemier** [comment]
You can add the following lines to get the base64-encoded secret as well:

```sh
base64secret="$(printf '%s' "$secret" | base64UrlEncode)" 
echo "$base64secret"
```

> Using that base64 value in jwt.io with the "secret base64 encoded" checkbox enabled now shows the same payload failing its signature verification.

That's because you encoded the secret with the end of line character added by `echo`. You can use `printf '%s'` to provide just the secret without the `\n` to `base64UrlEncode`:

```sh
$ printf '%s' 'abc123' | base64UrlEncode
YWJjMTIz
```

Then the signature checks on jwt.io with this plain text secret as well.

**Comment by eric-brechemier** [comment]
By printing a binary secret before and after the Data.String.Conversions, you can see that the value differs.

What is the conversion for?

**Comment by begriffs** [comment]
That's a good question. We have a function `jwtClaims` to get the claims out of a JWT as a map of names to values. It [one of](https://github.com/begriffs/postgrest/blob/master/src/PostgREST/Auth.hs#L69) its lines uses [decodeAndVerifySignature](https://hackage.haskell.org/package/jwt-0.6.0/docs/Web-JWT.html#v:decodeAndVerifySignature) to do the heavy lifting. This function is from the `jwt` package in haskell, and it was designed to take the secret as a custom type, [Secret](https://hackage.haskell.org/package/jwt-0.6.0/docs/Web-JWT.html#t:Secret). This type is a glorified `Text` value that prevents any part of our code from inspecting it after it has been created, and only code internal to the jwt module can look inside.

So far so good, but what confuses me is why the library chose Text as a representation rather than ByteString. Text holds UTF16 values, whereas ByteString is meant to faithfully hold a sequence of bytes. Maybe the library uses Text in case you want a secret that is comprised of words in human languages in a non Latin alphabet.

Now to our conversion. We obtain the secret from the command line using a library called `optparse-applicative`. In particular its `strOption` function, which gives us back a String. (This is a crappy type that should be purged from Haskell, it's a linked list of bytes.) Anyway like everywhere else in the PostgREST source code I use the `cs` function, short for [convertString](https://hackage.haskell.org/package/string-conversions-0.4/docs/Data-String-Conversions.html#v:convertString). It is defined to automatically call other more specific functions depending on the input and output types inferred by Haskell's type system. In our case it is going from `String` to `Text`, which if you look at the source code you see it does with [Data.Text.pack](https://hackage.haskell.org/package/text-1.2.2.1/docs/Data-Text.html#v:pack).

A comment about the pack function says it “performs replacement on invalid scalar values.” That sounds worrisome for us. I think we just need to find another way to convert the String, maybe going first to ByteString and thence through functions in [Data.Text.Encoding](https://hackage.haskell.org/package/text-1.2.2.1/docs/Data-Text-Encoding.html).

**Comment by begriffs** [comment]
I also created [an issue](https://bitbucket.org/ssaasen/haskell-jwt/issues/21/supporting-cryptographically-secure) on the jwt library about this.

/cc @juretta

**Comment by begriffs** [comment]
@eric-brechemier Stefan will be releasing v0.7.1 of the jwt package to allow us to use binary secrets. https://bitbucket.org/ssaasen/haskell-jwt/branch/issue/21-binary-secrets#diff

**Comment by eric-brechemier** [comment]
@begriffs Excellent news. Thanks for the follow-through.

**Comment by juretta** [comment]
jwt `0.7.1` is now available from Hackage: https://hackage.haskell.org/package/jwt

Sorry for the hassle.

**Comment by eric-brechemier** [comment]
Thanks a lot @juretta for fixing it right away!

**Comment by begriffs** [comment]
@eric-brechemier can you try the `jwt-binary` branch and see if it works for you? If it's good then I'll try to make a regression test that fails for master but passes for this branch.

**Comment by eric-brechemier** [comment]
@begriffs No, it is still not working. The validation status on jwt.io is still `invalid signature` unless the generated secret contains only valid UTF-8 characters.

Running `binarySecret . traceShowId . cs . traceShowId` shows that the binary secret is still corrupted by `cs`. For example:

```
"1\816\56509-DjH"
"1\204\176\237\178\189-DjH"
```

**Comment by juretta** [comment]
Please let me know if this is due to a problem with the `jwt` library so I can (if necessary) fix it up there.

**Comment by begriffs** [comment]
@eric-brechemier do you have an editor which supports finding the types of Haskell expressions? I'm curious if you put your cursor on the `cs` function in that expression and inquire about its type what it will say...if it is `String -> ByteString` in that instance.

**Comment by eric-brechemier** [comment]
@begriffs I am using vim. I am convinced that the `cs` conversion should be removed. But I do not think that `strOption` supports parsing to `ByteString` directly.

Removing `cs` results in:

```
Couldn't match type ‘[Char]’ with ‘BS.ByteString’
    Expected type: Parser BS.ByteString
      Actual type: Parser String
    In the second argument of ‘(<$>)’, namely
      ‘strOption
      (...)
```

**Comment by begriffs** [comment]
Well look at this, the `cs` function appears to be using a function from the utf8-string package:
https://github.com/soenkehahn/string-conversions/blob/master/src/Data/String/Conversions.hs#L97-L98

I had assumed it was doing `pack`. Try

```haskell
binarySecret . traceShowId . BS.pack . traceShowId
```

**Comment by eric-brechemier** [comment]
Using `binarySecret . traceShowId . BS.pack . traceShowId`:

```
"\56522a\56466\56457\56517\56522\56524\56530"
"\202a\146\137\197\202\204\210"
```

Using `binarySecret . traceShowId . cs . traceShowId`:

```
"\56522a\56466\56457\56517\56522\56524\56530"
"\237\179\138a\237\178\146\237\178\137\237\179\133\237\179\138\237\179\140\237\179\146"
```

Actual secret:

```sh
$ xxd jwt.secret 
0000000: ca61 9289 c5ca ccd2                      .a......
```

**Comment by eric-brechemier** [comment]
The conversion from hex to decimal gives:

* `ca` -> `\202`
* `61` -> `\97`
* `92` -> `\146`
* `89` -> `\137`
* `c5` -> `\197`
* `ca` -> `\202`
* `cc` -> `\204`
* `d2` -> `\210`

with  `\97` being the ASCII code of `a`, this looks good.

**Comment by eric-brechemier** [comment]
I tried a few different random keys with binary characters, the signatures generated by PostgREST using `binarySecret . traceShowId . BS.pack . traceShowId` now matches on jwt.io which returns 'Signature Verified'.

**Comment by begriffs** [comment]
Excellent! Can you test with secrets which use unicode? For instance 極秘.

**Comment by eric-brechemier** [comment]
The signature generated by PostgREST is also verified when using `極秘` as secret, when the secret is provided as tex to jwt.io.

The signature verification fails however when converting `極秘` to base64url as `5qW156eY` and providing this base64 value to jwt.io.

**Comment by eric-brechemier** [comment]
For a secret in ASCII (`abc` or `secret`), the verification succeeds both for text secret and secret converted to base64.

**Comment by eric-brechemier** [comment]
Same result with other Unicode secrets: `ⶆⶐ` or `∀™↻`: signature is declared valid when the secret is provided as text to jwt.io, invalid when provided as base64.

**Comment by eric-brechemier** [comment]
The signature verification on jwt.io is done with following code:

```js
window.verify = function (algorithm, value, key, isSecretBase64Encoded) {

  var result = '', error = null;

  if (algorithm === 'HS256'){
    if (isSecretBase64Encoded) {
      try {
        key = window.b64utob64(key);
        key = window.CryptoJS.enc.Base64.parse(key).toString();
      } catch (e) {
        return {result: '', error: e};
      }
    } else {
      key = window.CryptoJS.enc.Latin1.parse(key).toString();
    }
  }

  try {
    result = KJUR.jws.JWS.verify(value, key);
  } catch (e) {
    error = e;
  }

  return {result: result, error: error};
};
```

**Comment by eric-brechemier** [comment]
The functions `window.CryptoJS.enc.Base64.parse(key).toString()` and `window.CryptoJS.enc.Latin1.parse(key).toString()` convert the key to hex characters, in a format similar to xxd output:

```js
>> window.CryptoJS.enc.Base64.parse('VhGHbMPHgow').toString()
"5611876cc3c7828c"
```

```sh
$ xxd jwt.secret 
0000000: 5611 876c c3c7 828c                      V..l....
```

```js
>> window.CryptoJS.enc.Latin1.parse('secret').toString()
"736563726574"
```

```sh
$ printf 'secret' | xxd
0000000: 7365 6372 6574                           secret
```

**Comment by eric-brechemier** [comment]
In the case of `極秘`, the text signature actually corrupts the key by decoding it as Latin1:


```sh
$ printf '極秘' | xxd
0000000: e6a5 b5e7 a798                           ......
```

```js
>> window.CryptoJS.enc.Latin1.parse('極秘').toString()
"75d8"
```

while the base64 key matches the original:

```sh
$ printf '極秘' | base64UrlEncode
5qW156eY
```

```js
>> window.CryptoJS.enc.Base64.parse('5qW156eY').toString()
"e6a5b5e7a798"
```

**Comment by eric-brechemier** [comment]
In the case of `極秘`, the signature generated by `binarySecret . traceShowId . BS.pack . traceShowId` is actually invalid. It is only corrupted in the same way as providing `極秘` as Latin1 'text'. Keys with Unicode characters shall be provided encoded as base64 to jwt.io to avoid corruption.

On the other hand, for `極秘`, the signature generated by `binarySecret . traceShowId . cs . traceShowId` is valid; the key is not corrupted in this case:

```
"\26997\31192"
"\230\165\181\231\167\152"
```

**Comment by begriffs** [comment]
So what command line interface would you like? We could add a `--jwt-secret-base64` to deal with the random bytes and continue having `--jwt-secret` properly handle unicode.

**Comment by eric-brechemier** [comment]
I would prefer `--jwt-secret-path` actually, a path to the file which contains the secret. This would avoid any further confusion with encoding/decoding from base64 vs base64url etc.

**Comment by begriffs** [comment]
Would this run into the same problem since the file could contain 極秘 or could contain binary data that constitutes invalid unicode? Or does the file contain operating system metadata that can tell Haskell what the encoding is?

**Comment by eric-brechemier** [comment]
The difference is that you would not try to read this file as Unicode or plain text, and would only consider its binary contents. In the end, all the signature algorithms use bytes. The problem is to feed them the right bytes, unaltered.

**Comment by begriffs** [comment]
Got it, so whoever edits the file does so using an editor or command line tools that ensure the encoding is what they wanted it to be. Then we just read the bytes.

My next question is about breaking backward compatibility. What if we left the current command line flag as it is and use the `--jwt-secret "$(cat jwt.secret)"` trick? PR #582 will make it properly interpret the bytes, right?

**Comment by eric-brechemier** [comment]
No, because of the `cs` step between `binarySecret` and `strOption`.

**Comment by eric-brechemier** [comment]
We would need a `binaryOption` somehow.

**Comment by begriffs** [comment]
I see what you mean, sorry I was being dense. 😛 

I closed my pull request for this because it was insufficient. Marking this issue as something for 0.4 because of the backward incompatible parameter change.

**Comment by eric-brechemier** [comment]
What is the backward-incompatible change that you have in mind?

I suggest that we keep the current command line flag as it is and add a separate command line flag to load the secret from a file instead.

**Comment by begriffs** [comment]
Oh I thought it would be confusing to have both options. Since the file way always works but the string argument way sometimes fails I was planning to keep only the file way.

**Comment by eric-brechemier** [comment]
OK.

**Comment by eric-brechemier** [comment]
I am revisiting this issue in the context of PostgREST v0.4.x.

I confirm that it is now possible to use a random binary secret, if it is encoded in base64 and written inline in the configuration file:

```
jwt-secret = "jqHYd0C+Vjk="
secret-is-base64 = true
```

```
$ printf '%s' 'jqHYd0C+Vjk=' | base64 -D | xxd
00000000: 8ea1 d877 40be 5639                      ...w@.V9
```

I found two issues with secrets written in a file though, for which I created separate issues:

* [Binary secret cannot be read from a file #876](https://github.com/begriffs/postgrest/issues/876)
* [Base64 secret cannot be decoded from a file ending with a newline #877](https://github.com/begriffs/postgrest/issues/877)

**Comment by eric-brechemier** [comment]
@begriffs I backported secret reading from a file to v0.3.x, and fixed the file reading to support binary secrets. I added a separate parameter for the file path, to simplify.

Let me know if you are interested in these changes for v0.3.x. I can also help to fix the issue on v0.4.x. At the core, the fix is the use of `binarySecret <$> BS.readFile path` instead of `secret <$> readFile path`.

**Comment by begriffs** [comment]
Thanks! Most interested in the fix for 0.4, can you send a PR?

**Comment by eric-brechemier** [comment]
For future reference: this was done in commit e770570ce and merged with PR #988.
